### PR TITLE
Introduce shellcheck

### PR DESCRIPTION
**Pull request by szepeviktor**
It is strict and beneficial!

### CODE REVIEW

**Comment by szepeviktor** [comment]
@chriscool What do you think?
Does shellcheck move your project in a better direction?
https://travis-ci.org/chriscool/sharness/jobs/379455354

**Comment by chriscool** [comment]
@szepeviktor I would like to use shellcheck, but we need to fix at least the problems that make it fail the Travis CI build.

**Comment by szepeviktor** [comment]
Protecting `"$singles"` and `"${variables}with-strings`
Explicit `1>&2` redirs
etc.

**Comment by szepeviktor** [comment]
@chriscool Please do squash my commit mass.

File: sharness.sh [hunk 1]
```diff
@@ -614,10 +658,9 @@ test_cmp() {
 test_seq() {
 	i="$1"
 	j="$2"
-	while test "$i" -le "$j"
-	do
+	while test "$i" -le "$j"; do
 		echo "$i" || return
-		i=$(expr "$i" + 1)
+		i="$((i + 1))"
```
Comment (chriscool) [inline sharness.sh]: I think it would make Sharness less portable.

Comment (chriscool) [reply sharness.sh]: Actually Git uses Arithmetic Expansion $(( ... )), so it's ok to use it:

https://github.com/git/git/blob/master/Documentation/CodingGuidelines#L98

Sorry about the noise.

File: sharness.sh [hunk 2]
```diff
@@ -732,11 +772,11 @@ test_done() {
...
 
 	if test -z "$HARNESS_ACTIVE"; then
-		test_results_dir="$SHARNESS_TEST_DIRECTORY/test-results"
+		test_results_dir="${SHARNESS_TEST_DIRECTORY}/test-results"
 		mkdir -p "$test_results_dir"
-		test_results_path="$test_results_dir/$this_test.$$.counts"
+		test_results_path="${test_results_dir}/${this_test}.${$}.counts"
 
-		cat >>"$test_results_path" <<-EOF
+		cat >> "$test_results_path" <<-EOF
```
Comment (chriscool) [inline sharness.sh]: I don't think this is needed for shellcheck. And this goes against the Git coding guidelines:

https://github.com/git/git/blob/master/Documentation/CodingGuidelines#L57-L70

File: sharness.sh [hunk 3]
```diff
@@ -747,63 +787,65 @@ test_done() {
...
-		say "1..$test_count$skip_all"
+		say "1..${test_count}${skip_all}"
 
 		test_eval_ "$final_cleanup"
 
 		test -d "$remove_trash" &&
-		cd "$(dirname "$remove_trash")" &&
-		rm -rf "$(basename "$remove_trash")"
+		cd "$(dirname "$remove_trash")" \
+		&& rm -rf "$(basename "$remove_trash")"
```
Comment (chriscool) [inline sharness.sh]: This goes against the Git guidelines.

**Comment by chriscool** [comment]
In general it seems to me that there are a number of changes that are not necessary to pass shellcheck and I'd rather avoid them as they could make it more difficult to merge code from Git.

**Comment by szepeviktor** [comment]
OK. So should I close this PR?

**Comment by chriscool** [comment]
What I would suggest is:

- keep the first commit ("Introduce shellcheck") but add options `-e SC2034 -e SC2154 -e SC2094` to the shellcheck command to disable the checks that give false positives
- maybe create an issue to say that we should take a look at enabling those checks again later
- squash the rest of the commits into one commit changing sharness.sh and one changing aggregate-results.sh
- remove all the changes that are not necessary to pass shellcheck (with the options to disable checks that give false positives) from both commits
- move the first commit  ("Introduce shellcheck") after the other commits so that all the commits pass all the tests
- force push the branch so that this PR is updated

Thanks!

**Comment by chriscool** [comment]
@szepeviktor sorry for the long wait but I eventually took the time to look at this again, and I came up with #81

I think this new pull request about the same subject is just a start though. You (and others like @sumpfralle who worked on https://github.com/chriscool/sharness/pull/70) are welcome to improve step by step on that. I just prefer a step by step approach rather than trying to fix everything at once.

**Comment by chriscool** [comment]
For example I would welcome changes to `.travis.yml` like the one in your patches.

What do you think?

**Comment by chriscool** [comment]
Closed in favor of #81 
Thanks!

**Pull request closed without merging**